src/librustc_borrowck/borrowck/mir/dataflow/mod.rs

```diff
@@ -0,0 +1,1168 @@
+        let stmt = &mir.basic_block_data(bb).statements[idx];
+        let move_paths = &move_data.move_paths;
+        let loc_map = &move_data.loc_map;
+        let path_map = &move_data.path_map;
+        let rev_lookup = &move_data.rev_lookup;
+
+        let loc = Location { block: bb, index: idx };
+        debug!("stmt {:?} at loc {:?} moves out of move_indexes {:?}",
+               stmt, loc, &loc_map[loc]);
+        for move_index in &loc_map[loc] {
+            // Every path deinitialized by a *particular move*
+            // has corresponding bit, "gen'ed" (i.e. set)
+            // here, in dataflow vector
+            zero_to_one(&mut sets.gen_set, *move_index);
+        }
+        let bits_per_block = self.bits_per_block(ctxt);
+        match stmt.kind {
+            repr::StatementKind::Assign(ref lvalue, _) => {
+                // assigning into this `lvalue` kills all
+                // MoveOuts from it, and *also* all MoveOuts
+                // for children and associated fragment sets.
+                let move_path_index = rev_lookup.find(lvalue);
+
+                sets.kill_set.set_bit(move_path_index.idx());
+                on_all_children_bits(path_map,
+                                     move_paths,
+                                     move_path_index,
+                                     |moi| {
+                                         assert!(moi.idx() < bits_per_block);
+                                         sets.kill_set.set_bit(moi.idx());
+                                     });
+            }
+        }
+    }
+
+    fn terminator_effect(&self,
+                         ctxt: &Self::Ctxt,
+                         sets: &mut BlockSets,
+                         bb: repr::BasicBlock,
+                         statements_len: usize)
+    {
+        let &(_tcx, mir, ref move_data) = ctxt;
+        let term = mir.basic_block_data(bb).terminator.as_ref().unwrap();
+        let loc_map = &move_data.loc_map;
+        let loc = Location { block: bb, index: statements_len };
+        debug!("terminator {:?} at loc {:?} moves out of move_indexes {:?}",
+               term, loc, &loc_map[loc]);
+        let bits_per_block = self.bits_per_block(ctxt);
+        for move_index in &loc_map[loc] {
+            assert!(move_index.idx() < bits_per_block);
+            zero_to_one(&mut sets.gen_set, *move_index);
+        }
+    }
+
+    fn propagate_call_return(&self,
+                             ctxt: &Self::Ctxt,
+                             in_out: &mut [usize],
+                             _call_bb: repr::BasicBlock,
+                             _dest_bb: repr::BasicBlock,
+                             dest_lval: &repr::Lvalue) {
+        let move_data = &ctxt.2;
+        let move_path_index = move_data.rev_lookup.find(dest_lval);
+        let bits_per_block = self.bits_per_block(ctxt);
+
+        in_out.clear_bit(move_path_index.idx());
+        on_all_children_bits(&move_data.path_map,
```

Comment:
hurray for non-uniformity - consider extracting this as a function. Same bug as `statement_effect`.